src/mlir/generator.h

```diff
@@ -104,55 +104,39 @@ namespace mlir::verona
 
   /**
    * MLIR Generator.
-   *
-   * There are two entry points: AST and MLIR, and two exit points: MLIR
-   * and LLVM IR. The MLIR -> MLIR path is for self-validation and future
-   * optimisation passes (mainly testing).
-   *
-   * The LLVM IR can be dumped or used in the next step of the compilation,
-   * lowering to machine code by the LLVM library.
-   *
-   * For now, the error handling is crude and needs proper consideration,
-   * especially aggregating all errors and context before sending it back to
-   * the public API callers.
    */
   struct Generator
   {
-    Generator() : builder(&context)
-    {
-      // Opaque operations and types can only exist if we allow
-      // unregistered dialects to co-exist. Full conversions later will
-      // make sure we end up with onlt Verona dialect, then Standard
-      // dialects, then LLVM dialect, before converting to LLVM IR.
-      context.allowUnregisteredDialects();
+    /**
+     * Convert an AST into a high-level MLIR module.
+     *
+     * Currently this generates opaque MLIR operations, which can't be processed
+     * by the rest of the compiler, but we will slowly be transitioning to the
+     * Verona dialect.
+     *
+     */
+    static llvm::Expected<mlir::OwningModuleRef>
+    lower(MLIRContext* context, const ::ast::Ast& ast);
 
+  private:
+    Generator(MLIRContext* context) : context(context), builder(context)
+    {
       // Initialise known opaque types, for comparison.
       // TODO: Use Verona dialect types directly and isA<>.
-      allocaTy = genOpaqueType("alloca", context);
-      unkTy = genOpaqueType("unk", context);
-      noneTy = genOpaqueType("none", context);
+      allocaTy = genOpaqueType("alloca");
+      unkTy = genOpaqueType("unk");
+      noneTy = genOpaqueType("none");
       boolTy = builder.getI1Type();
     }
 
-    /// Read AST into MLIR module.
-    llvm::Error readAST(const ::ast::Ast& ast);
-    /// Read MLIR into MLIR module.
-    llvm::Error readMLIR(const std::string& filename);
-
-    /// Transform the opaque MLIR format into Verona dialect.
-    llvm::Error emitMLIR(const llvm::StringRef filename, unsigned optLevel = 0);
-    /// Transform the Verona dialect into LLVM IR.
-    llvm::Error emitLLVM(const llvm::StringRef filename, unsigned optLevel = 0);
-
     using Types = llvm::SmallVector<mlir::Type, 4>;
 
-  private:
     /// MLIR module.
     mlir::OwningModuleRef module;
+    /// MLIR context.
+    mlir::MLIRContext* context;
```

Comment:
Aren't you leaking this context, now that it's just a pointer and you don't destroy it?

Reply:
The context is owned by the driver (or whoever uses the generator). It's their responsibility to manage it.

I'll add a note to the constructor of Generator.

Reply:
Ah, makes sense!